### ISSUE TITLE
t-SNE

### ISSUE THREAD

**Issue by lanzagar** [biolab/orange3#3116]
A widget for t-SNE embedding and visualization should be part of the core Orange.
It is already implemented in one of the add-ons (single-cell), but is a general method that should be found next to the very similar MDS.

When moving the widget I propose we try to keep it as consistent as possible with the functionality and interface of MDS. That means we:
- remove the PCA Preprocessing section
- add Initialization choice
- after optimization finishes, pressing Start should run a new optimization from the current embedding (points can be reset by choosing one of the initializations again)
- I don't really see the Refresh option as necessary, but can be present as well if it makes the reuse of code between t-SNE and MDS easier (see below). Is it necessary/used in MDS?
- as the GUIs will be _very_ similar, we would ideally have both widgets reuse some common code to avoid copy pasting. We can a) move the widget to Orange first and then refactor the code in another PR or b) do it when introducing the widget in the first place

Open questions:
- the current t-SNE has some caching capabilities, do we want to remove that, introduce it to MDS too, or keep it just in t-SNE but not have it in MDS
- @pavlin-policar do you think any additional parameters would make sense when switching to your implementation or is Perplexity enough (along with max iter)?

**Comment by pavlin-policar** [comment]
I would include something that I think the current widget is lacking. Given that we use tSNE approximations, both Barnes-Hut and FI have a parameter that controls the trade-off between speed and quality (`theta`/`angle` for BH and `ints_in_inverval` for FI). With BH you don't really notice if the quality is bad, but with FI, you can observe some square-like formations in the embedding if the accuracy is too low. Having a slider from "Speed" to "Quality" makes a lot of sense to me that would control these parameters.

Another task would be to heuristically determine which approximation technique to use - BH is faster for data sets <10k while FI dominates for larger data sets. Same goes to determine when to switch from exact NN to approximate NN and to determine how many cores to use for small data sets.

`perplexity` and `max_iter` are definitely important, but it also might be useful to include some way to specify the early exaggeration phase, which needs to know how many iterations to run for (this is 250 by default and hard coded into the current implementation) and what the exaggeration factor should be (this is, again, currently hard coded and probably set to 12).

I guess we could leave these two hard-coded simply because then the user would be overwhelmed with the number of options and leave more advanced settings to be used in scripting.

A possible solution, but one that may cause some confusion, might be to add a slider for exaggeration [1, ) and then have the user specify that and optimize for e.g. 250 steps (i.e. the early exaggeration phase), set it to 1 and optimize for 750 steps (normal phase). Sometimes late exaggeration is used to better separate the clusters, so set exaggeration to > 1 then optimize for e.g. 100 steps. I'm just thinking aloud here, because this is probably far too complicated for the average user.

I have also found that sometimes changing the `learning_rate` parameter is useful for better embeddings, but usually, it's not an issue, so we probably want to hard code that as well.

**Comment by pavlin-policar** [comment]
I believe this is sufficiently implemented to close this.
